### PR TITLE
Extended semantic segmentation to image segmentation

### PR DESCRIPTION
**Pull request by merveenoyan**
This PR extends semantic segmentation guide to cover two other segmentation types (except for the big fine-tuning part) and compares them. cc @NielsRogge as discussed

### CODE REVIEW

File: docs/source/en/_toctree.yml
Comment (osanseviero) [inline docs/source/en/_toctree.yml:62]: This will break old urls. In case we really want to rename the file, you can set up a redirect by doing something like https://github.com/huggingface/hub-docs/blob/main/docs/hub/_redirects.yml

File: docs/source/en/tasks/semantic_segmentation.md
Comment (pcuenca) [inline docs/source/en/tasks/semantic_segmentation.md:152]: How can we see in this output that every pixel is classifed?

File: docs/source/en/tasks/semantic_segmentation.md
Comment (pcuenca) [inline docs/source/en/tasks/semantic_segmentation.md:188]: I'd maybe use the same order you used in the exposition: Reference, Semantic Segmentation, Instance Segmentation, Panoptic Segmentation.

The Instance Segmentation Output appears to contain more classes than "car" and "person", but the model output above didn't. Perhaps we could make it consistent?

File: docs/source/en/tasks/semantic_segmentation.md
Comment (amyeroberts) [inline docs/source/en/tasks/semantic_segmentation.md:23]: Similar to @pcuenca's note - it's not always objects that are being labelled depending on the type of segmentation happening.

```suggestion
Image segmentation models divide images into areas of interest such as different objects or background regions e.g. the sky. These models segment images by assigning a label to each pixel. There are three different types: semantic segmentation, instance segmentation, and panoptic segmentation.
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

+1 to all of @pcuenca's comments

File: docs/source/en/tasks/semantic_segmentation.md
Comment (MKhalusova) [inline docs/source/en/tasks/semantic_segmentation.md:1]: I don't think there's a need to update the copyright.

**Review by MKhalusova — COMMENTED** [review]
I agree with the comments from @pcuenca and @amyeroberts. We probably should also add a couple of headers. 
Right now, the right-side navigation looks like this:
```
## Load SceneParse150 dataset
## Preprocess
## Evaluate
## Train
## Inference
```

I would suggest the following structure:
```
## Types of segmentation
## Fine-tune a semantic segmentation model
### Load SceneParse150 dataset
### Preprocess
### Evaluate
### Train
### Inference
```

Also, in the inference example at the end of the fine-tuning section, we can probably leave only the example of doing inference manually, since we already show inference examples with a pipeline at the beginning of the doc.

File: docs/source/en/tasks/semantic_segmentation.md
Comment (merveenoyan) [reply docs/source/en/tasks/semantic_segmentation.md]: I fixed this sentence.

File: docs/source/en/tasks/semantic_segmentation.md
Comment (merveenoyan) [reply docs/source/en/tasks/semantic_segmentation.md:189]: Surprisingly that building is classified as car, and this is one of the best (maybe it is the best) instance segmentation models on Hub (mask2former). I'd rather not modify?

**Comment by merveenoyan** [comment]
I addressed all the comments. Sorry I deprioritized it shortly.

File: docs/source/en/tasks/semantic_segmentation.md
Comment (amyeroberts) [inline docs/source/en/tasks/semantic_segmentation.md:23]: It's not always objects 

```suggestion
Image segmentation models separate areas corresponding to different areas of interest in an image. These models work by assigning a label to each pixel. There are several types of segmentation: semantic segmentation, instance segmentation, and panoptic segmentation.
```

**Review by amyeroberts — APPROVED** [review]
Thanks! Just one small nit

**Comment by merveenoyan** [comment]
I see an error in CI yet I don't understand it, seems like it's about something else?

File: docs/source/en/tasks/semantic_segmentation.md
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.md:46]:
```suggestion
We can use Transformers' image segmentation pipeline to quickly infer with a semantic segmentation model called [SegFormer](model_doc/segformer). Let's take a look at the example image.
```
Not sure the link here will work

File: docs/source/en/tasks/semantic_segmentation.md
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.md:104]:
```suggestion
Taking a look at the mask for the car class, we can see every car is classified with the same mask.
```

File: docs/source/en/tasks/semantic_segmentation.md
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.md:113]:
```suggestion
In instance segmentation, the goal is not to classify every pixel, but to predict a mask for **every instance of a class** in a given image. We will use [facebook/mask2former-swin-large-cityscapes-instance](https://huggingface.co/facebook/mask2former-swin-large-cityscapes-instance) for this.
```
I would add here that instance segmentation is very similar to object detection: you want to get a set of instances out of your image, the only difference between object detection and instance segmentation being that the former gets a bounding box per instance, whereas the latter gets a binary mask per instance

File: docs/source/en/tasks/semantic_segmentation.md
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.md:146]:
```suggestion
Panoptic segmentation combines semantic segmentation and instance segmentation, where every pixel is classified, and there are multiple masks for each instance of a class. We can use [facebook/mask2former-swin-large-cityscapes-panoptic](https://huggingface.co/facebook/mask2former-swin-large-cityscapes-panoptic) for this.
```
Panoptic segmentation technically assigns 2 labels per pixel: a semantic category and an instance ID.

File: docs/source/en/tasks/semantic_segmentation.md
Comment (MKhalusova) [reply docs/source/en/tasks/semantic_segmentation.md:113]: That's a really nice way to build intuition for instance segmentation!

File: docs/source/en/tasks/semantic_segmentation.md
Comment (MKhalusova) [inline docs/source/en/tasks/semantic_segmentation.md:41]: This paragraph is repeated later in the fine-tuning section. It's probably best to have this information only once. I would suggest to leave the library installation instructions here (as we need the libraries installed for inference examples to work), but the notebook log in makes more sense in the fine-tuning section.

**Review by MKhalusova — APPROVED** [review]
Thanks for iterating on the guide! LGTM, only one minor nit - the notebook login and pip install section is repeated twice in the guide.

File: docs/source/en/tasks/semantic_segmentation.md
Comment (merveenoyan) [reply docs/source/en/tasks/semantic_segmentation.md]: Thanks! I addressed it

**Comment by merveenoyan** [comment]
@MKhalusova can you merch if you approve? 👉 👈  🥹

File: docs/source/en/tasks/semantic_segmentation.md
Comment (pcuenca) [inline docs/source/en/tasks/semantic_segmentation.md:1]: Maybe rename file/URL to `image_segmentation.md`, for consistency with the contents. (Also in the yaml, of course)  :)

File: docs/source/en/tasks/semantic_segmentation.md
Comment (pcuenca) [reply docs/source/en/tasks/semantic_segmentation.md:46]: I think it would

File: docs/source/en/tasks/semantic_segmentation.md
Comment (merveenoyan) [reply docs/source/en/tasks/semantic_segmentation.md:1]: Done!

Comment (merveenoyan) [reply docs/source/en/tasks/semantic_segmentation.md:1]: Done!

**Comment by merveenoyan** [comment]
doc-builder is still looking for `semantic_segmentation.md` although I've put a redirection 🧐 
Screenshot 2023-11-20 at 17 42 47

**Comment by MKhalusova** [comment]
We need green CI checks before we can merge. It looks like the `utils/check_doctest_list.py` check is failing. You can run it locally to debug and fix.

**Comment by pcuenca** [comment]
@MKhalusova it's my fault, I suggested to rename the file for consistency. I opened [this PR](https://github.com/merveenoyan/transformers/pull/1/files) to @merveenoyan's repo, which fixes the problem locally. I also fixed an issue with `check_task_guides.py`. Please, let me know if there's a better way to deal with this kind of situation :)

I also saw a `semantic_segmentation.md` in the `ko` translation, but refrained from renaming it because there's no `_redirect.yml` for Korean. Should we add it?

**Comment by MKhalusova** [comment]
@pcuenca Generally, we tend to avoid renaming/removing files. I have no experience safely doing so, perhaps @stevhliu could advise?

**Comment by pcuenca** [comment]
It's not a big deal, we can just go back to @merveenoyan's version before the rename if that's simpler.

**Comment by mishig25** [comment]
Regarding redirects. For example, `perf_infer_gpu_many: perf_infer_gpu_one`. There is a chicken-egg situation because hf.co/docs uses [_redirects.yml that is on main currently](https://github.com/huggingface/transformers/blob/main/docs/source/en/_redirects.yml). Unless, this PR gets merged hf.co/docs for now will not redirect`perf_infer_gpu_many: perf_infer_gpu_one`

**Comment by merveenoyan** [comment]
@MKhalusova according to Mishig's response we need to merge before it turns red and then it will be green, so maybe you can make the call in this case.

**Comment by pcuenca** [comment]
There are two different things here.

- The redirection works for me in the docs generated for the PR, these both work:
https://moon-ci-docs.huggingface.co/docs/transformers/pr_27039/en/tasks/semantic_segmentation (old)
https://moon-ci-docs.huggingface.co/docs/transformers/pr_27039/en/tasks/image_segmentation (new)

- There are other CI scripts that fail because of references to the old name, as shown in this PR: https://github.com/merveenoyan/transformers/pull/1/files
These tests won't become green if we merge.

Given the increased complexity and that @MKhalusova said we generally try to avoid renames, I'd suggest we remove the rename and keep the same filename it had before. Sorry for introducing noise!

**Comment by merveenoyan** [comment]
Can this be merged by someone with write access?

**Comment by amyeroberts** [comment]
I can merge it :)

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating!

**Pull request merged**
